juju_verify/verifiers/ceph.py

```diff
@@ -109,60 +140,101 @@ def __hash__(self) -> int:
         """Return hash representation of AZ objects."""
         return hash(self.__str__())
 
-    def find_parent(self, node_id: int) -> Optional[NodeInfo]:
-        """Find the node that has the id in the children list."""
-        for node in self._nodes:
-            if node.children and node_id in node.children:
-                return node
+    def _get_node(self, name: str) -> NodeInfo:
+        """Get node by name."""
+        if name not in self._nodes_name_map.keys():
+            raise KeyError(f"Node {name} was not found.")
+
+        try:
+            node = self._nodes[self._nodes_name_map[name]]
+            assert node.name == name  # check that variable _nodes was not changed
+            return node
+        except (IndexError, AssertionError) as error:
+            raise ValueError("Private value `_nodes` was changed.") from error
+
+    def _find_ancestor(self, node: NodeInfo, required_type: str) -> Optional[NodeInfo]:
+        """Find ancestor w/ the desired type.
+
+        This function will recursively search for the parent node until the parent
+        is of the desired type.
+        Example:
+            [
+                {"id": -1, "name": "root", "children": [-2, -3], ...},
+                {"id": -2, "name": "rack.0", "children": [-4, -5], ...},
+                {"id": -4, "name": "host.0", "children": [0, 1, 2], ...},
+                {"id": -5, "name": "host.1", "children": [3, 4, 5], ...},
+                ...
+                {"id": -3, "name": "rack.1", "children": [-6, -7], ...},
+                ...
+            ]
+            The request is to find the `root` ancestor for the `host.0`.
+            The first step is to find a parent who has id=-4 among its children, then
+            check if it is root. The parent node found is of the rack type with id=-2,
+            so the first step is repeated for this node until the parent node is of the
+            root type.
+        """
+        for _node in self._nodes:
+            if _node.children and node.id in _node.children:
+                if _node.type != required_type:
+                    # continue searching for the parent for the currently found node
+                    return self._find_ancestor(_node, required_type)
+
+                return _node
 
         return None
 
-    def get_nodes(self, name: str) -> List[NodeInfo]:
-        """Get nodes by their name."""
-        nodes = [node for node in self._nodes if node.name == name]
-        if not nodes:
-            raise KeyError(f"Node {name} was not found.")
+    def can_remove_host_node(
+        self, *names: str, required_ancestor_type: str = "root"
+    ) -> bool:
+        """Check if host node could be removed."""
+        if required_ancestor_type not in self.SUPPORTED_ANCESTOR_TYPES:
+            raise ValueError(f"`{required_ancestor_type}` is not supported")
 
-        return nodes
+        if not all(self._get_node(name).type == "host" for name in names):
```

Comment:
Add a comment saying that: "names allowed are of the type "host", which matches Juju units".